DRILL-5318: Sub-operator test fixture

paul-rogers opened the pull request:
This commit depends on:

* DRILL-5319
* DRILL-5323
* DRILL-5324

This PR cannot be accepted (or built) until the above are pulled and
this PR is rebased on top of them. The PR is issued now so that reviews
can be done in parallel.

Provides the following:

* A new OperatorFixture to set up all the objects needed to test at the
sub-operator level. This relies on the refactoring to create the
required interfaces.
* Pulls the config builder code out of the cluster fixture builder so
that configs can be build for sub-operator tests.
* Modifies the QueryBuilder test tool to run a query and get back one
of the new row set objects to allow direct inspection of data returned
from a query.
* Modifies the cluster fixture to create a JDBC connection to the test
cluster. (Use requires putting the Drill JDBC project on the test class
path since exec does not depend on JDBC.)

sohami commented on exec/java-exec/src/test/java/org/apache/drill/test/QueryBuilder.java:
```diff
@@ -271,6 +276,91 @@ public QuerySummary run() throws Exception {
+    }
+    RowSetReader reader = rowSet.reader();
+    reader.next();
+    long value = reader.column(0).getLong();
```
Can we please refactor the common portion in singletonLong and singletonInt ? Since I guess in future if we support retrieving more types then again it will be repeated?

sohami commented on exec/java-exec/src/test/java/org/apache/drill/test/QueryBuilder.java:
```diff
@@ -271,6 +276,91 @@ public QuerySummary run() throws Exception {
+    QueryDataBatch dataBatch = null;
+    for (QueryDataBatch batch : results()) {
+      if (dataBatch == null  &&  batch.getHeader().getRowCount() != 0) {
+        dataBatch = batch;
```
Ok makes sense after reading this: _"The reason for the funny code is that we want to release memory for all but the first batch."_

sohami commented:
Thanks for the changes. LGTM. +1

paul-rogers commented on exec/java-exec/src/test/java/org/apache/drill/test/ConfigBuilder.java:
```diff
@@ -0,0 +1,130 @@
+    }
+  }
+
+  private Properties configProperties() {
```
Note that this has all been replaced by your recent addition...

sudheeshkatkam commented:
+1

sudheeshkatkam commented:
This does not compile. Please fix the errors and squash the commits into one.

paul-rogers commented:
Rebased onto DRILL-5318 so that this branch includes the DRILL-5318 code that this PR depends upon, as well as the additional code required for this PR. Fixed the merge errors that caused the compilation issue.

Please commit DRILL-5318 first. Then, commit this PR. Given that this PR is already rebased on 5318, this PR should merge cleanly into master.

sudheeshkatkam commented:
I am confused by your comment. Which should be committed first, 5318 or 5323? Your comments says 5318, but the commit message for 5318 and the fact that 5318 is based on 5323 suggests 5323 should be committed first?

Please format the commit message in 70fdc88 so it starts with "DRILL-XXXX: ..."

paul-rogers commented:
Please see the initial comment which identifies that this PR depends on several others. Two have been done, only DRILL-5323 remains. Please commit DRILL-5323 first, then this one.

Note that, per your request, this PR is rebased onto DRILL-5323. The history here includes all changes from DRILL-5323 plus additional changes specific to DRILL-5318. As I understand it, if you first commit DRILL-5323 the common changes will reside in master, and will be ignored when you then (as a second step) commit this PR.

paul-rogers commented:
Revised the commit header. Note that the commit messages also identifies that this PR is dependent on DRILL-5323, and so this PR must be committed after DRILL-5323.

Sorry for the confusion. The work was broken into multiple commits to ease the burden on code reviewers. Unfortunately, there is no free lunch, and the multiple commits puts a large burden on the submitter to keep all branches in sync, and on the committer to pull the PRs in the correct order.

jinfengni commented on exec/java-exec/src/test/java/org/apache/drill/test/rowSet/test/RowSetTest.java:
```diff
@@ -0,0 +1,400 @@
+
+  @AfterClass
+  public static void tearDownAfterClass() throws Exception {
+    fixture.close();
```
Is this the only unit test added to DRILL-5318 / DRILL-5323 which are for unit test framework related enhancements? Looks like the new tests in this class are to test the new added concept such as "RowSet", "RowSetSchema", "FlattenedSchema". Do we have some example that shows how to use those concepts to do a operator / sub-operator testing, which I thought are the main purpose of the two PRs?

paul-rogers commented:
One of the key reasons for creating tests is that they show how to use an API. In this case, the associated tests show how to use the row set abstractions to create a schema, populate a row set and so on. Please also see the package-info.java file for Javadoc that provides an overview.

Please let me know if you want to use the test framework. If so, I'd like to work with you to figure out what additional documentation and/or examples would be useful.

This PR is a dependency on the "main show": a PR with refactored sort code and complete sort unit tests. That PR will be submitted this week now that all the prerequisites are in master. That will demonstrate now to use the framework to create a suite of unit tests. I can point to my personal repo if if it would be helpful to preview that code.

jinfengni commented:
It's fine to have example in a follow-up PR.

I was trying to see if I can use new operator/sub-operator test framework for the schema change task I'm doing. That requires to work with one or multiple RecordBatch (both data and operator).
